### ISSUE TITLE
False positive on rule RF02 when using DECLARE / SET in BigQuery

### ISSUE THREAD

**Issue by Noloxs** [sqlfluff/sqlfluff#6191]
### Search before asking

- [X] I searched the [issues](https://github.com/sqlfluff/sqlfluff/issues) and found no similar issues.

### What Happened

I have a BigQuery query defined which starts with: DECLARE two variables then using SET.
However when running lint afterwards it marks it with a RF02 error about qualifying the reference, when I attempt to use it. 

### Expected Behaviour

I expect it to ignore such declared variables as they cannot be qualified further

### Observed Behaviour

The linting marks the declared variable as needing further reference qualification.

### How to reproduce

I have simplified and anonymized the  query to one that shows the issue:

Query:
```
DECLARE period_format String;

SET period_format = '%Y-%Q'; # The date format used for the given period

SELECT
    current_period.event_date,
    current_period.account_id,
    current_period.peer_id,
    FORMAT_DATE(period_format, DATE(current_period.event_date)) AS period,
    COUNT(DISTINCT current_period.event_date) OVER (PARTITION BY TRUE) AS days_in_period
FROM
    `dateset.table` AS current_period
LEFT JOIN
    {{ data_table }} AS previous_periods
    USING
        (account_id)
WHERE
    FORMAT_DATE(period_format, DATE(current_period.event_date))
    = '2024-3'
```
Config:
```
[sqlfluff]
dialect = bigquery
templater = jinja
# Ignore rule about replacing USING https://docs.sqlfluff.com/en/stable/reference/rules.html#rule-ST07
exclude_rules = ST07 

[sqlfluff:templater:jinja:context]
data_table = `dataset.table2`

[sqlfluff:rules:references.special_chars]
allow_space_in_identifier = True
```
When running:
```
sqlfluff lint --config sqlfluff_bigquery <file>
```

I get the following result:
```
FAIL                                                                                                 
L:   9 | P:  17 | RF02 | Unqualified reference 'period_format' found in select                                                
                       | with more than one referenced table/view.
                       | [references.qualification]
L:  10 | P:   5 | LT05 | Line is too long (88 > 80). [layout.long_lines]
L:  14 | P:  25 | AL05 | Alias 'previous_periods' is never used in SELECT
                       | statement. [aliasing.unused]
L:  18 | P:  17 | RF02 | Unqualified reference 'period_format' found in select
                       | with more than one referenced table/view.
                       | [references.qualification]
```

This is were I would expect it to not trigger on line: 9 and 18

### Dialect

bigquery

### Version

sqlfluff, version 3.1.1
Python 3.12.5

### Configuration

[sqlfluff]
dialect = bigquery
templater = jinja
# Ignore rule about replacing USING https://docs.sqlfluff.com/en/stable/reference/rules.html#rule-ST07
exclude_rules = ST07 

[sqlfluff:templater:jinja:context]
data_table = `test`

[sqlfluff:rules:references.special_chars]
allow_space_in_identifier = True

### Are you willing to work on and submit a PR to address the issue?

- [ ] Yes I am willing to submit a PR!

### Code of Conduct

- [X] I agree to follow this project's [Code of Conduct](https://github.com/sqlfluff/sqlfluff/blob/main/CODE_OF_CONDUCT.md)

**Comment by keraion** [comment]
Duplicate of #2991
